grouper/models.py

```diff
@@ -1406,17 +1406,76 @@ def apply_changes(self, request):
+                    if expiration > datetime.utcnow():
+                        AsyncNotification.add_expiration(self.session,
+                                                         expiration,
+                                                         group_name,
+                                                         member_name,
+                                                         recipients=recipients,
+                                                         member_is_user=member_is_user)
                 else:
                     setattr(self, key, None)
             else:
                 setattr(self, key, value)
 
+    def expire(self):
+        """Mark a single edge as inactive and log to the audit log."""
+        # TODO(rra): Arbitrarily use the first listed owner of the group from
+        # which membership expired as the actor, since we have to provide an actor
+        # and we didn't record who set the expiration on the edge originally.
+        actor_id = next(self.group.my_owners().itervalues()).id
+
+        # Pull data about the edge and the affected user or group.
+        group_name = self.group.name
+        if OBJ_TYPES_IDX[self.member_type] == "User":
+            user = User.get(self.session, pk=self.member_pk)
+            member_name = user.username
+            recipients = [member_name]
+            member_is_user = True
+        else:
+            subgroup = Group.get(self.session, pk=self.member_pk)
+            member_name = subgroup.groupname
+            recipients = subgroup.my_owners_as_strings()
+            member_is_user = False
+
+        # Log to the audit log.  How depends on whether a user's membership has
+        # expired or a group's membership has expired.
+        audit_data = {
+            "action": "expired_from_group",
+            "actor_id": actor_id,
+            "description": "{} expired out of the group".format(member_name),
+        }
+        if member_is_user:
+            AuditLog.log(
+                self.session, on_user_id=user.id, on_group_id=self.group_id,
+                **audit_data
+            )
+        else:
+            # Make an audit log entry for both the subgroup and the parent group
+            # so that it will show up in the FE view for both groups.
+            AuditLog.log(self.session, on_group_id=self.group_id, **audit_data)
+            AuditLog.log(self.session, on_group_id=subgroup.id, **audit_data)
+
+        # Send email notification.
+        email_context = {
+            "group_name": group_name,
+            "member_name": member_name,
+            "member_is_user": member_is_user,
+        }
+        from grouper.fe.settings import settings as fe_settings
+        send_email(
+            session=self.session,
+            recipients=recipients,
+            subject="Membership in {} expired".format(group_name),
+            template="expiration",
+            settings=fe_settings,
+            context=email_context,
+        )
```

Comment:
i suspect we'll want to remove `send_email` and `async_send_email` dependencies within models. this will get us out of the circular dependency hole we're in now.

it's a little more annoying but sending the email would be cleaner in `audit.expire_edges` and will be one less dependency to remove when the time comes.

for the `async_send_email` dependency i was thinking of passing a callback into the model method. that could work here though it's simple enough you might be able to detach it entirely. your call.